advanced_topics.qmd

```diff
@@ -5,19 +5,220 @@
 library(arrow)
 library(dplyr)
 ```
-
 ```{r}
 #| label: setup-advanced-data-manip
 #| include: false
 pums_person <- open_dataset("./data/person")
 ```
 
+In this chapter we're going to dive into some more advanced topics.
+As we've shown in the previous chapters, the arrow R package includes support for hundreds of functions in R and has a rich type system.
+These empower you to solve most common data analysis problems.
+However, if you ever need to extend that set of functionality, there are several options.
+
+To start, we'll look in more detail at how bindings from R functions to Acero, the Arrow C++ query engine, work.
+We'll also explore the converse of this: functions that exist in Acero and don't have R equivalents, and how you can access these in your arrow pipelines.
+Then, we're going to show how you can call back from Acero into R to do calculations where functions you want to use inside don't have a corresponding Acero function.
+For other cases, we'll show how the Arrow format's focus on interoperability lets you stream data to and from other query engines, such as DuckDB, to access different functionality they have.
+Finally, we'll look at a practical case of extending Arrow's type system to work with geospatial data.
+
+## Acero
+
+Acero is the Arrow C++ library's query execution engine.
+A query execution engine is a library which allows users to perform queries against a data source.
+It takes input data and the operations to be performed on the data, combines them, and returns the output from this process.
+
+It typically has multiple steps, which could include things like:
+
+* taking the query and parsing it into an algebraic format
+* re-ordering and optimising the query in order to run it in the most efficient way[^optimizer]
+* actually running the query
+
+[^optimizer]: Acero doesn't have a logical query optimizer, but as we have seen, it does support optimizations like predicated pushdown: this is what makes filtering on partitioned datasets fast. To take advantage of these optimzations, there is code in the R package that reorders queries before passing it to Acero for most common workflows.
```

Comment:
```suggestion
[^optimizer]: Acero doesn't have a logical query optimizer, but as we have seen, it does support optimizations like "predicate pushdown" - filtering the data where it is stored rather than when in memory, which is what makes filtering on partitioned datasets fast. To take advantage of these optimizations, there is code in the R package that reorders queries before passing it to Acero for most common workflows.
```